query-engine/core/src/executor/mod.rs

```diff
@@ -131,3 +132,53 @@ pub trait TransactionManager {
 pub fn get_current_dispatcher() -> Dispatch {
     tracing::dispatcher::get_default(|current| current.clone())
 }
+
+#[cfg(not(target_arch = "wasm32"))]
+pub(crate) mod task {
+    use super::*;
+
+    pub type JoinHandle<T> = tokio::task::JoinHandle<T>;
+
+    pub fn spawn<T>(future: T) -> JoinHandle<T::Output>
+    where
+        T: Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        tokio::spawn(future)
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+pub(crate) mod task {
+    use super::*;
+
+    #[pin_project::pin_project]
+    pub struct JoinHandle<T>(#[pin] tokio::sync::oneshot::Receiver<T>);
```

Comment:
`task::JoinHandle<T>` is a Wasm-compatible alternative to `tokio::task::JoinHandle<T>`.

It uses the `pin_project` macro to enable [pin-projection](https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning).

Reply:
This might be related to my previous comment, actually: https://github.com/prisma/prisma-engines/pull/4445#discussion_r1396949303

Can you provide me with links to the documentation as of how you realized this was needed?

Reply:
Yes. So, this is not Wasm-specific. It's rather related to the implementation of a `Future` trait from scratch for a `tokio::sync::oneshot::Receiver` (i.e., a single-use received end of a synchronized tokio channel.

Consider:

```rust
use tokio::sync::oneshot::{self};

pub struct JoinHandle<T>(oneshot::Receiver<T>);

impl<T> Future for JoinHandle<T> {
    type Output = Result<T, oneshot::error::RecvError>;

    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
        todo!()
    }
}
```

Let's try replacing the `todo!`:

```rust
self.0.poll(cx)
```

yields

```txt
error[E0599]: no method named `poll` found for struct `tokio::sync::oneshot::Receiver` in the current scope
   --> 
    |
37  |             self.0.poll(cx)
    |                    ^^^^ method not found in `Receiver<T>`

    |
105 |     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
    |        ---- the method is available for `Pin<&mut tokio::sync::oneshot::Receiver<T>>` here
    |
help: consider wrapping the receiver expression with the appropriate type
    |
37  |             Pin::new(&mut self.0).poll(cx)
    |             +++++++++++++       +
```

Let's apply the suggestion:

```rust
core::pin::Pin::new(&mut self.0).poll(cx)
```

which however yields

```txt
error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
  --> 
   |
37 |             core::pin::Pin::new(&mut self.0).poll(cx)
   |                                      ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
35 |         fn poll(mut self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
   |                 +++
```

Applying the new suggestion, we get:

```rust
```rust
use tokio::sync::oneshot::{self};

pub struct JoinHandle<T>(oneshot::Receiver<T>);

impl<T> Future for JoinHandle<T> {
    type Output = Result<T, oneshot::error::RecvError>;

    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
        std::pin::Pin::new(&mut self.0).poll(cx)
    }
}
```

which compiles, but leaves us exposed to the risk of pins being used "the wrong way".

`pin-project` is a utility commonly used when dealing with pin requirements around `futures` (there's a cool [blog post](https://blog.yoshuawuyts.com/safe-pin-projections-through-view-types/) on the matter).
There's also a similar, slimmer crate, [`pin-project-lite`](https://github.com/taiki-e/pin-project-lite), which could be worth taking a look at.